pkg/nacler/keyers/secureenclave/secureenclave.go

```diff
@@ -0,0 +1,194 @@
+//go:build darwin
+// +build darwin
+
+package secureenclave
+
+// this was heavily inspiried by https://github.com/facebookincubator/sks
+// thank you!
+
+/*
+#cgo darwin LDFLAGS: -framework Foundation -framework Security -framework CoreFoundation
+
+#include <stdlib.h>
+#include <secureenclave.h>
+
+typedef struct wrapper {
+	unsigned char *buf;
+	int status;
+	size_t size;
+	char *error;
+} Wrapper;
+
+Wrapper *wrapCreateKey() {
+	Wrapper *res = (Wrapper *)malloc(sizeof(Wrapper));
+	if (!res)
+		return NULL;
+	memset(res, 0, sizeof(Wrapper));
+	res->size = createKey(&res->buf, &res->error);
+	return res;
+}
+
+Wrapper *wrapFindKey(void *hash) {
+	Wrapper *res = (Wrapper *)malloc(sizeof(Wrapper));
+	if (!res)
+		return NULL;
+	memset(res, 0, sizeof(Wrapper));
+	res->size = findKey((unsigned char *)hash, &res->buf, &res->error);
+	return res;
+}
+
+Wrapper *wrapECDH(void *hash, void *counterParty, int counterPartySize) {
+	Wrapper *res = (Wrapper *)malloc(sizeof(Wrapper));
+	if (!res)
+		return NULL;
+	memset(res, 0, sizeof(Wrapper));
+	res->size = ecdh((unsigned char *)hash, (unsigned char *)counterParty, counterPartySize, &res->buf, &res->error);
+	return res;
+}
+*/
+import "C"
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/sha1"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"unsafe"
+)
+
+type SecureEnclaveKeyerOption func(*SecureEnclaveKeyer)
+
+func WithExistingKey(publicKey ecdsa.PublicKey) SecureEnclaveKeyerOption {
+	return func(s *SecureEnclaveKeyer) {
+		s.publicKey = &publicKey
+	}
+}
+
+type SecureEnclaveKeyer struct {
+	publicKey *ecdsa.PublicKey
+}
+
+func New(opts ...SecureEnclaveKeyerOption) (*SecureEnclaveKeyer, error) {
+	se := &SecureEnclaveKeyer{}
+
+	for _, opt := range opts {
+		opt(se)
+	}
+
+	// if the call provided a public key, make sure we can find it in secure enclave
+	if se.publicKey != nil {
+		_, err := findKey(*se.publicKey)
+		if err != nil {
+			return nil, fmt.Errorf("finding existing public key: %w", err)
+		}
+
+		return se, nil
+	}
```

Comment:
This (and the associated `WithExistingKey`) makes sense, but I wonder if we want a different signature. As I understand it, I think _most_ of the usage should be with an existing generated public key. So I wonder about making the public key a mandatory argument, and exposing a separate `NewPublicKey(...)`

I'm thinking about how this is likely to be called:
```go
key := bucket.Get(...)
if key == nil {
  key = CreateKey(...)
  bucket.Put(key)
}

keyer := New(key)
```

vs needing to pull the key, vary the arguments, and thread back a saving the key

Reply:
yea, that is cleaner, done